modules/frontend/v1/request_batch.go

```diff
@@ -102,13 +110,17 @@ func (b *requestBatch) doneChan(stop <-chan struct{}) <-chan struct{} {
 
 // reportErrorToPipeline sends errors back up the query frontend http pipeline
 func (b *requestBatch) reportErrorToPipeline(err error) {
+	b.mx.RLock()
```

Comment:
Is it ok to block on the chan writes while holding this lock?  Could that lead to deadlock?
At least the hierarchy should be documented.